modules/magniteAnalyticsAdapter.js

```diff
@@ -0,0 +1,887 @@
+    const gamHasRendered = deepAccess(cache, `auctions.${auction.auctionId}.gamRenders.${adUnit.adUnitCode}`);
+    return adUnit.adUnitCode === bidWonData.adUnitCode && gamHasRendered;
+  }
+  let { adUnit, auction } = findMatchingAdUnitFromAuctions(matchingFunction, false);
+  // If no match was found, we will use the actual bid won auction id
+  return {
+    renderTransactionId: (adUnit && adUnit.transactionId) || bidWonData.transactionId,
+    renderAuctionId: (auction && auction.auctionId) || bidWonData.auctionId
+  }
+}
+
+const formatBidWon = bidWonData => {
+  // get transaction and auction id of where this "rendered"
+  const { renderTransactionId, renderAuctionId } = getRenderingIds(bidWonData);
+
+  const isCachedBid = renderTransactionId !== bidWonData.transactionId;
+  logInfo(`${MODULE_NAME}: Bid Won : `, {
+    isCachedBid,
+    renderAuctionId,
+    renderTransactionId,
+    sourceAuctionId: bidWonData.auctionId,
+    sourceTransactionId: bidWonData.transactionId,
+  });
+
+  // get the bid from the source auction id
+  let bid = deepAccess(cache, `auctions.${bidWonData.auctionId}.auction.adUnits.${bidWonData.adUnitCode}.bids.${bidWonData.requestId}`);
+  let adUnit = deepAccess(cache, `auctions.${bidWonData.auctionId}.auction.adUnits.${bidWonData.adUnitCode}`);
+  let bidWon = {
+    ...bid,
+    sourceAuctionId: bidWonData.auctionId,
+    renderAuctionId,
+    transactionId: bidWonData.transactionId,
+    sourceTransactionId: bidWonData.transactionId,
+    bidId: bid.pbsBidId || bidWonData.bidId || bidWonData.requestId, // if PBS had us overwrite bidId, use that as signal
+    renderTransactionId,
+    accountId,
+    siteId: adUnit.siteId,
+    zoneId: adUnit.zoneId,
+    mediaTypes: adUnit.mediaTypes,
+    adUnitCode: adUnit.adUnitCode,
+    isCachedBid: isCachedBid || undefined // only send if it is true (save some space)
+  }
+  delete bidWon.pbsBidId; // if pbsBidId is there delete it (no need to pass it)
+  return bidWon;
+}
+
+const formatGamEvent = (slotEvent, adUnit, auction) => {
+  const gamEvent = pick(slotEvent, [
+    // these come in as `null` from Gpt, which when stringified does not get removed
+    // so set explicitly to undefined when not a number
+    'advertiserId', advertiserId => isNumber(advertiserId) ? advertiserId : undefined,
+    'creativeId', creativeId => isNumber(slotEvent.sourceAgnosticCreativeId) ? slotEvent.sourceAgnosticCreativeId : isNumber(creativeId) ? creativeId : undefined,
+    'lineItemId', lineItemId => isNumber(slotEvent.sourceAgnosticLineItemId) ? slotEvent.sourceAgnosticLineItemId : isNumber(lineItemId) ? lineItemId : undefined,
+    'adSlot', () => slotEvent.slot.getAdUnitPath(),
+    'isSlotEmpty', () => slotEvent.isEmpty || undefined
+  ]);
+  gamEvent.auctionId = auction.auctionId;
+  gamEvent.transactionId = adUnit.transactionId;
+  return gamEvent;
+}
+
+const subscribeToGamSlots = () => {
+  window.googletag.pubads().addEventListener('slotRenderEnded', event => {
```

Comment:
why not use slotResponseReceived

Reply:
I actually do not know! This is what we have been using in the old Rubicon Analytics Adapter.

What are the differences between the two? I will go look at documentation but curious as to your thoughts.

Reply:
Looks like `slotResponseReceived` does not have the gam ID's included. At least from the documentation standpoint.

Reply:
it does though, they are just undocumented

Reply:
also, annoyingly, yieldgroupid and companyid are not documented

Reply:
does it really?

I looked and it did not pass it, will double check, if so the slotResponseReceived is def better as it fires earlier!

Reply:
@patmmccann On my test page I get the following:

![image](https://user-images.githubusercontent.com/13972794/195411315-11e147a4-014c-44bc-849c-6b0b2e368eb7.png)